### PR TITLE
✨ Set Custom Referrer Fees

### PR DESCRIPTION
**Pull request by easyrun42**
### Description

Update! For your "lock", set fees like 20% for "vitalik.eth" or 50% for "david.eth". Just head to lock -> advanced -> referral fee. Adjust and see all referrers there. Easy!

Issue: [Show and set custom referrer fees](https://github.com/unlock-protocol/unlock/issues/12557)

### Demo (Unlock-app & Subgraph)
Quick demo showing how it works for both existing "locks" and for creating a brand new "lock".

### Existing Lock

https://github.com/unlock-protocol/unlock/assets/136458308/4c352fb9-64d6-43a3-9280-96a919270d9b

### Brand New Lock

https://github.com/unlock-protocol/unlock/assets/136458308/d2d1efb6-e4c4-4d01-9a50-5099747ecf64

### Subgraph (Goerli & Optimism)

Optimism: https://thegraph.com/hosted-service/subgraph/easyrun2023/unlock-optimism-mainnet

Goerli: https://thegraph.com/hosted-service/subgraph/easyrun2023/unlock-goerli

https://github.com/unlock-protocol/unlock/assets/136458308/0e85ca44-2c52-4b8c-a3a7-f62850af500c

### CODE REVIEW

File: .clabot [hunk 1]
```diff
@@ -48,7 +48,9 @@
     "christopherdro",
     "patrickworkman",
     "PJijin",
-    "adeelhasan"
+    "adeelhasan",
+    "KosmosKey",
```
Comment (easyrun42) [inline .clabot:52]: Added myself as an contributor. @KosmosKey did the commit, but pull request was raised from @easyrun42 👍

File: packages/unlock-js/src/web3Service.ts [hunk 2]
```diff
@@ -905,18 +905,35 @@ export default class Web3Service extends UnlockService {
...
+      lock(id:"${lockAddress}") {
+        referrerFees {
+          id
+          referrer
+          fee
+        }
+      }
+    }`
+
+    const result = await fetch(
```
Comment (easyrun42) [inline packages/unlock-js/src/web3Service.ts:922]: The `referrerFees` function retrieves all `referrerFees` from a "lock" address.

This code will be updated once the graph is uploaded to the original "Unlock Protocol" subgraph on all networks. Afterwards, I will pull in the graph URLs based on the network using import networks from `@unlock-protocol/networks` ✅.

File: packages/unlock-js/src/web3Service.ts [hunk 3]
```diff
@@ -905,18 +905,35 @@ export default class Web3Service extends UnlockService {
...
+          'Content-Type': 'application/json',
+          Accept: 'application/json',
+        },
+        body: JSON.stringify({ query }),
+      }
     )
-    const referrerFees = await lockContract.referrerFees(address)
-    return ethers.BigNumber.from(referrerFees).toNumber()
+      .then((r) => r.json())
+      .then((res) => res?.data?.lock?.referrerFees || [])
```
Comment (easyrun42) [inline packages/unlock-js/src/web3Service.ts:934]: Getting the referrer fees and in case it's `undefined` then we will receive an empty array

File: subgraph/schema.graphql [hunk 4]
```diff
@@ -33,6 +33,8 @@ type Lock @entity {
   deployer: Bytes!
   "Total number of receipts of lock"
   numberOfReceipts: BigInt!
+  "Referrers fees set per custom address"
+  referrerFees: [ReferrerFee!]! @derivedFrom(field: "lock")
```
Comment (easyrun42) [inline subgraph/schema.graphql:37]: We've added `referrerFees` to the Lock entity to track the referrer fees set for each address.

**Improvement Tip** _(if it doesn't align with the product stack)_: Keep using `referrerFees` in the Lock but switch to a format like `${referrer}-${fee}`, for instance `0x123...-200`. This array of strings can be processed with `.split("-")` in the browser to extract the address and fee percentage.

Comment (easyrun42) [reply subgraph/schema.graphql:37]: @julien51 Let me know if this still suits you 👍

File: subgraph/schema.graphql [hunk 5]
```diff
@@ -117,3 +119,14 @@ type Receipt @entity {
   "Increasing number of receipt"
   receiptNumber: BigInt!
 }
+
+type ReferrerFee @entity {
```
Comment (easyrun42) [inline subgraph/schema.graphql:123]: Here is the `ReferrerFee` entity created 🙌

File: subgraph/src/public-lock.ts [hunk 6]
```diff
@@ -399,3 +406,16 @@ export function handleLockMetadata(event: LockMetadataEvent): void {
     lock.save()
   }
 }
+
+export function handleReferrerFees(event: ReferrerFeeEvent): void {
```
Comment (easyrun42) [inline subgraph/src/public-lock.ts:410]: This function tracks all referrer fee events for your "lock". If the "lock" exists, it logs the referrer fee set for an address. This way, we can monitor the "fee", "referrer", and associated "lock".

File: subgraph/subgraph.yaml [hunk 7]
```diff
@@ -109,5 +110,8 @@ templates:
         - event: LockMetadata(string,string,string)
           handler: handleLockMetadata
           receipt: true
+        - event: ReferrerFee(indexed address,uint256)
```
Comment (easyrun42) [inline subgraph/subgraph.yaml:113]: Here is the event we will be listening to in all of the evm blockchains that `Unlock Protocol` supports

File: unlock-app/src/components/interface/locks/Settings/elements/SettingMisc.tsx [hunk 8]
```diff
@@ -52,7 +52,7 @@ export const SettingMisc = ({
     <div className="grid grid-cols-1 gap-6">
       <SettingCard
         label="Referral fee"
-        description="Set up a percentage of the membership price to be sent to the referrer. This is great use cases to reward your members to promote your membership."
+        description="Set custom referral percentage for each address. Easily decide how much each promoter earns from your membership."
```
Comment (easyrun42) [inline unlock-app/src/components/interface/locks/Settings/elements/SettingMisc.tsx:55]: Updated the description because the feature now supports adding a referrer fee for a particular address 🙌.

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 9]
```diff
@@ -16,82 +17,90 @@ interface UpdateReferralFeeProps {
...
     })
   }
 
   const setReferrerFeeMutation = useMutation(setReferrerFee)
 
   const onSubmit = async (fields: FormProps) => {
-    const setReferrerFeePromise = setReferrerFeeMutation.mutateAsync(fields)
+    const setReferrerFeePromise = setReferrerFeeMutation.mutateAsync({
+      ...fields,
+      referralAddress: ethers.utils.getAddress(
```
Comment (easyrun42) [inline unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:71]: We wanna make sure the address is an check summed address otherwise smart contract might throw an error ❌

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 10]
```diff
@@ -16,82 +17,90 @@ interface UpdateReferralFeeProps {
...
+        fields.referralAddress.toLowerCase()
+      ),
+    })
 
     await ToastHelper.promise(setReferrerFeePromise, {
-      loading: 'Updating referrer fee',
+      loading: 'Setting new referrer',
       error: 'Failed to update the values, please try again.',
-      success: 'Referrer fee updated.',
+      success: 'New referrer set',
```
Comment (easyrun42) [inline unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:79]: Minor updates

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 11]
```diff
@@ -16,82 +17,90 @@ interface UpdateReferralFeeProps {
...
   )
 
-  useEffect(() => {
-    setValue('referralFeePercentage', (referralFeePercentage ?? 0) / 100)
-    setReferralFee((referralFeePercentage ?? 0) > 0)
-  }, [referralFeePercentage])
+  const referralFees = referralFeesData || []
 
-  const disabledInput =
+  const isDisabledReferrerInput =
```
Comment (easyrun42) [inline unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:92]: Ensure that variable naming is clear for other developers to understand. It should always start with "is", "has", or "should".

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 12]
```diff
@@ -105,19 +114,78 @@ export const UpdateReferralFee = ({
...
-          disabled={disabledInput}
+          disabled={isDisabledReferrerInput}
           loading={setReferrerFeeMutation.isLoading}
         >
           Apply
         </Button>
       )}
+
+      <div className="flex flex-col mt-5 items-start">
+        <Card.Title>Referrers ({referralFees.length})</Card.Title>
```
Comment (easyrun42) [inline unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:157]: You will have a complete overview of the referrers set per address on the lock you created ✅.

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 13]
```diff
@@ -16,82 +17,90 @@ interface UpdateReferralFeeProps {
...
-      address: ZERO,
+      address: fields?.referralAddress,
       feeBasisPoint: fields?.referralFeePercentage * 100,
     })
   }
 
-  const getReferrerFees = async () => {
-    return await web3Service.referrerFees({
+  const getReferrerFees = async (): Promise<ReferrerFee[]> => {
+    return web3Service.referrerFees({
```
Comment (easyrun42) [inline unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:60]: I've been also using return await... but it's better to just use return... since both methods returns a promise 👍.

**Comment by easyrun42** [comment]
@julien51 Please let me know if there are any changes or adjustments you'd like to see in this PR. 👍

**Comment by easyrun42** [comment]
I use emojis from [gitmoji](https://gitmoji.dev/) for my commits. It helps quickly indicate the purpose of each change. Check their site to decode the emojis.

File: subgraph/tests/referrer.test.ts [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+  lockAddress,
+  lockManagers,
+  nullAddress,
+  referrerFee,
+} from './constants'
+import { Address, BigInt, Bytes } from '@graphprotocol/graph-ts'
+import { handleReferrerFees } from '../src/public-lock'
+import { Lock } from '../generated/schema'
+
+describe('Referrer', () => {
```
Comment (easyrun42) [inline subgraph/tests/referrer.test.ts:22]: Added some unit test for the Referrer ✅ ✨

**Review by searchableguy — APPROVED** [review]
Aside from one minor nit, LGTM

**Comment by searchableguy** [comment]
Could you make a separate pull request adding yourself as a contributor? Both of your accounts.

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 15]
```diff
@@ -105,19 +114,78 @@ export const UpdateReferralFee = ({
...
+        <Input
+          type="text"
+          {...register('referralAddress', {
+            validate: (value) =>
+              ethers.utils.isAddress(value.toLowerCase()) || 'Invalid address',
+            required: {
+              value: true,
+              message: 'This field is required.',
+            },
+          })}
```
[the remark this answers is not in the document]
Comment (easyrun42) [reply unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx]: Sounds good 👍. I have finally resolved ENS name here too: [Commit](https://github.com/easyrun42/unlock/commit/a3bbbad32c75264c3aeef9b7221d40a513130abd)

You can now add address or ens name 🙌

File: packages/unlock-js/src/web3Service.ts [hunk 16]
```diff
@@ -905,18 +905,35 @@ export default class Web3Service extends UnlockService {
...
+    const result = await fetch(
+      'https://api.thegraph.com/subgraphs/name/easyrun2023/unlock-goerli',
+      {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+          Accept: 'application/json',
+        },
+        body: JSON.stringify({ query }),
+      }
```
[the remark this answers is not in the document]
Comment (easyrun42) [reply packages/unlock-js/src/web3Service.ts]: Fixed here too 👍: [Commit](https://github.com/easyrun42/unlock/commit/a3bbbad32c75264c3aeef9b7221d40a513130abd)

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 17]
```diff
@@ -16,82 +18,115 @@ interface UpdateReferralFeeProps {
...
+    ['getLock', lockAddress, network, setReferrerFeeMutation.isSuccess],
+    async () => getLock()
   )
 
-  useEffect(() => {
-    setValue('referralFeePercentage', (referralFeePercentage ?? 0) / 100)
-    setReferralFee((referralFeePercentage ?? 0) > 0)
-  }, [referralFeePercentage])
+  // @ts-ignore
+  const referralFees = (referralFeesData?.referrer as ReferrerFee[]) || []
```
Comment (easyrun42) [inline unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:116]: It's not good to use `@ts-ignore` but i do this for now since when i try to build my `subgraph` fields it has this error:
 
referrer_thegraph

The easiest way @julien51 is if you can update the subgraph on all networks so i can rebuild it on my end

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 18]
```diff
@@ -16,82 +18,115 @@ interface UpdateReferralFeeProps {
...
     })
   }
 
-  const getReferrerFees = async () => {
-    return await web3Service.referrerFees({
-      lockAddress,
-      network,
-      address: ZERO,
-    })
+  const getLock = async () => {
```
Comment (easyrun42) [inline unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:60]: Here we will get the `lock` and then the lock object will have the `referrer` array

**Comment by easyrun42** [comment]
@searchableguy I've added this account to .clabot, but the other account needs a PR review check: https://github.com/unlock-protocol/unlock/pull/12581/

**Comment by searchableguy** [comment]
@easyrun42 merged

**Comment by easyrun42** [comment]
@searchableguy Quick question about subgraph. If i were to update something on the subgraph do i need to wait until you, julien or the team updates the production subgraph so i can get the latest results ?

Asking since test are failing right now 🙂. Let me know what you guys do.

**Comment by searchableguy** [comment]
@easyrun42 subgraph tests shouldn't rely on live subgraph, I believe since @clemsos made the changes.

File: packages/unlock-js/src/web3Service.ts [hunk 19]
```diff
@@ -899,26 +899,6 @@ export default class Web3Service extends UnlockService {
     return await unlockContract.publicLockLatestVersion()
   }
 
-  /**
-   * Returns referrer fees
-   */
-  async referrerFees({
```
Comment (julien51) [inline packages/unlock-js/src/web3Service.ts:905]: Hum, why did you remove this?

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx [hunk 20]
```diff
@@ -105,19 +133,76 @@ export const UpdateReferralFee = ({
...
+      <div className="grid gap-2">
+        <ToggleSwitch
+          title="Referrer address"
+          enabled={isReferralAddressEnabled}
+          disabled={isDisabledReferrerInput}
+          setEnabled={(enabled: boolean) => {
+            setIsReferralAddressEnabled(enabled)
+          }}
+        />
+        <Input
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx:149]: Please use our AddressInput component so that it does ENS resolution.

File: packages/unlock-js/src/web3Service.ts (hunk 19 above)
Comment (easyrun42) [reply packages/unlock-js/src/web3Service.ts:905]: Good question.

Before we used to get referrer fees from the smart contract:

image

With this update we no longer check referrer fees via smart contract. Instead we use subgraph 😁

File: unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx (hunk 20 above)
Comment (easyrun42) [reply unlock-app/src/components/interface/locks/Settings/forms/UpdateReferralFee.tsx]: Updated 👍 [🧼 Resolve ENS name with <AddressInput /> component](https://github.com/unlock-protocol/unlock/pull/12574/commits/1daac8f8281271eb402a3e0dc4ad168fcfc133a2)

File: packages/unlock-js/src/subgraph/schema.graphql [hunk 21]
```diff
@@ -25,6 +26,26 @@ query allLocks(
...
+        tokenAddress
+        symbol
+        price
+        lockManagers
+        version
+        maxNumberOfKeys
+        maxKeysPerAddress
+        createdAtBlock
+        totalKeys
+      }
```
Comment (julien51) [inline packages/unlock-js/src/subgraph/schema.graphql:47]: I am not sure we need to add the lock details here again since `referrerFees` is already _inside_ the `lock` object.

File: packages/unlock-js/src/web3Service.ts (hunk 19 above)
Comment (julien51) [reply packages/unlock-js/src/web3Service.ts:905]: Let's still keep it in the lib please as someone may want to use it without using the subgraph.

File: subgraph/schema.graphql (hunk 4 above)
Comment (julien51) [reply subgraph/schema.graphql:37]: It does!

File: subgraph/schema.graphql [hunk 22]
```diff
@@ -117,3 +119,14 @@ type Receipt @entity {
...
+  "Transaction Hash"
```
Comment (julien51) [inline subgraph/schema.graphql:124]: This is not right? Please update the comment :)

**Review by julien51 — CHANGES_REQUESTED** [review]
This looks good from a code perspective!

That said I would love to do a few things:
1. split the PR between the subgraph changes and the UI changes so we can deploy the subgraph without breaking the UI :)
2. We're missing the ability to set a referrer fee for _any_ address (by using address `0x0`), as it does currently.

I think we can fix 2 by changing the UI a bit. First I would remove the toggle next to the fee itself. then I would change a the "address" part to be disabled unless a toggle with "custom address" is set.

Does it make sense?

File: subgraph/schema.graphql (hunk 22 above)
Comment (easyrun42) [reply subgraph/schema.graphql:124]: Yeah i think that one should be right since i use "Transaction Hash" as `ID!`

**Comment by easyrun42** [comment]
Subgraph PR here: https://github.com/unlock-protocol/unlock/pull/12596

Will close this one for now. The frontend PR will come right after subgraph pr has been merged
